internal/faustjs.org/docs/gutenberg/tutorial/create-a-custom-block.mdx

```diff
@@ -0,0 +1,196 @@
+
+<a href="/docs/img/creating-wp-blocks-folder.png" target="_blank">
+  <img
+    src="/docs/img/creating-wp-blocks-folder.png"
+    alt="The file structure of the project, showing the newly created wp-blocks folder at the root level of the Faust Scaffold app and containing a newly created folder called MyCoolBlock.js."
+  />
+</a>
+
+```jsx title="wp-blocks/MyCoolBlock.js"
+export default function MyCoolBlock(props) {
+  console.log(props.attributes);
+  return <div>MyCoolBlock</div>
+}
+```
+
+This component prints the block attributes via the props passed as a parameter. This will help us understand what is happening a little later on.
+
+### 4.2 Export the block in the `index.js`
+
+Make another file in `wp-blocks`, called `index.js`, and place the following inside:
+
+```jsx title="wp-blocks/index.js"
+import MyCoolBlock from './MyCoolBlock';
+export default {
+  MyCoolBlock
+}
+```
+
+By exporting the block in the `index.js`, we make it available in the `WordPressBlocksProvider`, which gives the proper context (i.e., which blocks have been defined as apart of your Faust app) to the `WordPressBlocksViewer` component. Read more about the WordPressBlocksProvider [here](http://localhost:3000/docs/reference/WordPressBlocksProvider). 
+
+### 4.3 Create the block fragment
+
+Next, you want to create a fragment that describes the request block fields and attributes. Ideally you want to include all the relevant attributes that this backend block function will need to call to your frontend block:
+
+```jsx title="wp-blocks/MyCoolBlock.js" {1}
+import {gql} from '@apollo/client';
+...
+MyCoolBlock.fragments = {
+  key: `MyCoolBlockBlockFragment`,
+  entry: gql`
+    fragment MyCoolBlockBlockFragment on MyCoolBlock {
+      attributes {
+        color
+        content
+        message
+      }
+    }
+  `,
+};
+```
+We chose to attach the fragment as a property of the function component itself. This follows best practice for creating [colocated fragments](https://www.apollographql.com/docs/react/data/fragments/#creating-colocated-fragments).
+
+### 4.4 Include the block fragment into the page query string
+
+You want to include the fragment in the page query string and reference the block fragment key so that when the page resolves, the payload includes the data from that block.
+
+In this example we will use it in the single page template provided in the Faust Scaffold. You can learn more about the Faust Template Hierarchy [here](/docs/templates).
+
+```jsx {1,7,18} title="wp-templates/single.js"
+import blocks from '../wp-blocks';
+export default function Component(props) {
+...
+}
+Component.query = gql`
+...
+${blocks.MyCoolBlock.fragments.entry}
+query GetPost(
+...
+	post(id: $databaseId, idType: DATABASE_ID, asPreview: $asPreview) {
+		...
+		contentBlocks {
+			name
+			__typename
+			renderedHtml
+			id: nodeId
+			parentId
+			...${blocks.MyCoolBlock.fragments.key}
+      	}
+	}
+...
+}
+```
```

Comment:
`ReferenceError: blocks is not defined` has halted progress on this item. I think it has to do with the blocks package, but I have installed `npm i @faustwp/blocks` and am not seeing where the import for `_app.js`should be importing from (`import { WordPressBlocksProvider } from '@faustwp/blocks';` is referenced in the https://canary.faustjs.org/docs/gutenberg/getting-started tutorial, but I'm not sure where it is in the repo)

Reply:
Here is the repo that corresponds with this tutorial testing: https://github.com/TeresaGobble/MERL-742-Tutorial

Reply:
It has to do with this line:

https://github.com/TeresaGobble/MERL-742-Tutorial/blob/main/pages/_app.js#L7

The config expects a `blocks` property but you are providing a `MyCoolBlock` property. This should have been:

`import blocks from '../wp-blocks'`

Reply:
Thank you Theo! I don't have any function that is `blocks` coming from my wp-blocks folder, though, so I still get the error that `ReferenceError: blocks is not defined` even with the original `blocks` import. I had changed it because I have the `MyCoolBlock` as the export in that folder and wasn't sure if it needed to be changed to match. Did I miss a step somewhere to add `blocks` to the `wp-blocks` folder in my practice repo?

Reply:
This is the full error: 
<img width="1299" alt="image" src="https://user-images.githubusercontent.com/50935135/221615469-494537c8-acb9-49f6-acc0-c8fb8c1450a4.png">

Reply:
This is all that I have in my `wp-blocks` folder so far! 
<img width="300" alt="image" src="https://user-images.githubusercontent.com/50935135/221616113-d01374ac-bdde-418e-bf29-cea74c6cccb7.png">

Reply:
Hi @theodesp! I paired with @blakewilson today and we found two additional main issues. In addition to your fix above and in the single template, I had two other missing pieces: the block was named differently than anticipated by the boiler plate from WordPress, so I was calling `MyCoolBlock` when I should have been using `CreateBlockMycoolblock`. I've added a step in the tutorial to address this. I also needed to install `composer` under the `wp-graphql-content-blocks-main` plugin folder (I could have sworn I had- Blake realized that I had no vender folder in there- I'll be sure to take note of this too!) 

We talked about adding a ticket for troubleshooting, since part of the challenge was being unable to access the full error message that Apollo was sending back.

Thanks again to both of you for your help! I'll keep you updated with my progress!

Reply:
CreateBlockMycoolblock is the `__typename` field of the block. It would also work with `name` as well.
This is documented here:
https://faustjs.org/docs/gutenberg/tutorial/create-a-block-from-wordpress-core#52-export-the-block-in-the-indexjs